File: PascalTriangle/PascalTriangle.cpp
#include<stdio.h>
#include<stdlib.h>

typedef struct node {
	int data;
	struct node * next;
}node;

typedef struct queue {
	node * front;
	node * rear;
	int size;
}Queue;



void InitQueue(Queue* Q) {
	node* q = (node *)malloc(sizeof(node));
	if (!q) {
		printf("stackoverflow");
		exit(0);
	}
	Q->front = q;
	Q->rear = q;
	Q->size = 0;
}

void DestroyQueue(Queue* Q) {
	if (Q != NULL)
		Q = NULL;
}

void ClearQueue(Queue* Q) {
	if (Q->front != Q->rear)
		Q->rear = Q->front;
}

bool QueueEmpty(Queue* Q) {
	return (Q->front == Q->rear);
}

void GetHead(Queue Q, int* e) {
	*e = Q.front->data;
}

void EnQueue(Queue * Q, int e) {
	node* q = (node *)malloc(sizeof(node));
	if (!q) {
		printf("stackoverflow");
		exit(0);
	}
	Q->rear->data = e;
	Q->rear=Q->rear->next = q;
	Q->size++;
}

void DeQueue(Queue* Q, int* e) {
	if (!QueueEmpty(Q)) {
		*e = Q->front->data;
		Q->front = Q->front->next;
		Q->size--;
	}
}
int QueueLength(Queue Q) {
	return Q.size;
}

int main() {
	Queue Triangle;
	InitQueue(&Triangle);
	EnQueue(&Triangle,1);
	EnQueue(&Triangle, 2);
	EnQueue(&Triangle, 1);

	int a, b, i = 0;
	int foobar;

	while(i <11) {
		a = Triangle.front->data;
		node* foo = Triangle.front->next;
		b = foo->data;
		DeQueue(&Triangle, &foobar);
		printf("%5d", foobar);
		if (a == 1) {
			EnQueue(&Triangle, 1);
			if (b == 1) {
				printf("\n");
				++i;
			}
			else
				EnQueue(&Triangle,a+b);
		}
		else {
			EnQueue(&Triangle, a + b);
		}
	}
	return 0;
}

File: PascalTriangle/PascalTriangle.java
//利用队列打印杨辉三角

package PascalTriangle;

public class Queue {
    class node {
        int data;
        node next;
    }

    private node fount;
    private node rear;
    private int length;

    {
        fount = new node();
        rear = new node();
        length = 0;
        fount = rear;
    }

    public void setFount (node fount) {
        this.fount = fount;
    }

    public void setRear (node rear) {
        this.rear = rear;
    }

    public node getFount () {
        return fount;
    }

    public node getRear () {
        return rear;
    }

    public void setLength (int length) {
        this.length = length;
    }

    public int getLength () {
        return length;
    }


    public boolean QueueEmpty () {
        if (fount == rear) {
            return true;
        }
        return false;
    }

    public void EnQueue (int e) {
        rear.data = e;
        rear.next = new node();
        rear = rear.next;
        length++;
    }

    public int DeQueue () {
        if (!QueueEmpty()) {
            int e = fount.data;
            fount = fount.next;
            length--;
            return e;
        }
        System.out.println("没有数据");
        System.exit(0);
        return 0;
    }

    //返回队列长度，不如getLength()好用
    public int QueueLength () {
        int length = 0;
        node foo = fount;
        while (foo.next != null) {
            ++length;
            foo = foo.next;
        }
        return length;
    }

    public static void main (String[] args) {
        Queue Triangle = new Queue();
        Triangle.EnQueue(1);
        Triangle.EnQueue(2);
        Triangle.EnQueue(1);
        int a, b, i = 0;
        while (i < 11) {
            a = Triangle.getFount().data;
            b = Triangle.getFount().next.data;
            System.out.print(Triangle.DeQueue() + " ");
            if (a == 1) {
                Triangle.EnQueue(1);
                if (b == 1) {
                    System.out.println();
                    ++i;
                } else
                    Triangle.EnQueue(a + b);
            } else {
                Triangle.EnQueue(a + b);
            }
        }
    }
}
